### PR TITLE
feat!: [SIW-453] Credential Issuing flow

### PR DESCRIPTION
**Pull request by balanza**
This PR adds a general flow for credential issuing. 
In order to do it, it introduces a **huge** refactor of both issuing and presentation flow. Below is an overview of changes to help readers to navigate through file diffs.

### List of Changes

**Package API**
The API exposed by the package has been reorganized using the following namespaces:
* `Credential.Issuance` contains the functions and types that map the steps of the issuing flow
* `Credential.Presentation` contains the functions and types that map the steps of the presentation flow
* `Trust` contains the functions and types that interact with the trusted entities

```ts
import { Credential, Trust } from "@pagopa/io-react-native-wallet"

Credential.Issuance.obtainCredential(...)
Credential.Presentation.getRequestObject(...)
Trust.getEntityStatement(...)
```
see `src/index.ts`

**Removed modules**
The modules `src/pid` and `src/rp` have been removed. Their implementations have been moved into `credential` as they are now generalized. `src/pid/sw-jwt` is still there as it implements the PID model.

**Code organization**
Being that both Issuance and Presentation flows are processes that concatenate input and output from a series of functions,
* One file for every step exposing the type that defines the input and output of the step, and its implementation
  * see files in `src/credential/issuance` and `src/credential/presentation`
* Each file is named with a number so that they are listed in the same order as they appear in the flow* 
* Steps that do not have a concrete implementation still have a type definition, so that the user application can develop it's own implementation; the rationale is that some operations require User interactions through UI elements that are out of scope for this package
  * see  `src/credential/issuance/01-start-flow.ts` as an actual example
* Function signatures are designed with the same shape:
  * they take as input the output of previous steps and the user-provided data
  * the last argument is a `context` object that maps what the step need from the environment (keys, attestations, configurations, etc)
  * they always return `Promise<Record<string, T>>`, even when only one value is returned; the rationale is to have a more eloquent return type
  * they throw in the operation fails

**utils: `Out<FN>`**
Just an alias for `Awaited<ReturnType<FN>>`, useful to link one function's output as another's input.
Rationale: make connections through steps more evident.

```ts
import { type Out } from "src/utils/misc"

declare function fn1(): Promise<{ foo: string }>
declare functon fn2(foo: Out<typeof fn1>["foo"]): Promise<void>
```

**utils: `hasStatus(number)`**
Simple helper that takes a `Response` object and throws an exception if it's not the expected status.
Rationale: code reuse and standardization

```ts
// before
const response = await appFetch("<url>")

if (response.status === 200) {
  const payload = await response.json();
  return { foo: payload.foo, bar: payload.bar };
}
throw new CustomError(
  `Unable to do something. Response code: ${await response.text()}`
);

// after
import { hasStatus } from "src/utils/misc"

return appFetch("<url>")
  .then(hasStatus(200)) // <-- a standard Http Error will be thrown
  .then(res => res.json())
  .then(payload => ({ foo: payload.foo, bar: payload.bar });
```

**Things fixed on the way**
* Add `credential_definition` for the supported credential metadata in `src/trust/types.ts`
* Add optional `wallet_relying_party` metadata for Credential issuer in `src/trust/types.ts`
* Make credential `type` generic in SD-JWT schema in `src/sd-jwt/types.ts`

**Workarounds**
* ~When obtaining Credential, before returning the received token we validate it. The signature validation fails for now, it's maybe an error in the Issuer? For now, we just log the error and skip the validation (see https://github.com/pagopa/io-react-native-wallet/blob/3772a5210e563cfdea44f7c3fc3dd27fd33b4e37/src/credential/issuance/06-obtain-credential.ts#L115)~ fixed
* Ignore missing `response_code` in https://github.com/pagopa/io-react-native-wallet/blob/e21362908b7b4ba31ef28dded4b55b8116242526/src/credential/presentation/04-send-authorization-response.ts#L18
* Normalize PID Issuer response with stub data so that is compliant to Credential issuance specification in https://github.com/pagopa/io-react-native-wallet/blob/dd1048632bac5348cf31c77e8962b0152e50c6ff/src/credential/issuance/03-start-user-authorization.ts#L39

### How Has This Been Tested?
Example app

### Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

----
Rewrite of #54

### CODE REVIEW

**Comment by balanza** [comment]
Open points:
* are we fine with step naming?
* are we fine with the proposed API?

File: src/credential/issuance/04-complete-user-authorization.ts
Comment (grausof) [inline src/credential/issuance/04-complete-user-authorization.ts:28]: Why is this specification not yet implemented? Could you detail it better? Maybe we also add a TODO with a reference task

File: src/credential/issuance/06-obtain-credential.ts
Comment (grausof) [inline src/credential/issuance/06-obtain-credential.ts:26]: TODO?

File: src/credential/issuance/06-obtain-credential.ts
Comment (grausof) [inline src/credential/issuance/06-obtain-credential.ts:122]: ⚠️ I think this is not acceptable. The problem must be fixed by the issuer if it exists

File: src/credential/presentation/03-get-request-object.ts
Comment (grausof) [inline src/credential/presentation/03-get-request-object.ts:79]:
```suggestion
  // Ensure that the request object conforms to the expected specification.
```

File: src/credential/presentation/04-send-authorization-response.ts
Comment (grausof) [inline src/credential/presentation/04-send-authorization-response.ts:26]:
```suggestion
 * Choose an RSA public key from those offered by the RP for encryption.
```

File: src/credential/presentation/04-send-authorization-response.ts
Comment (grausof) [inline src/credential/presentation/04-send-authorization-response.ts:55]:
```suggestion
/**
 * Generate a Verified Presentation token for a received request object within the context of an authorization request flow.
 * The presentation is created by revealing data from the provided credentials based on the requested claims.
 * Each Verified Credential is accompanied by the claims that the user consents to disclose from it.
 *
 * @todo: Allow for handling more than one Verified Credential.
 */
```

File: src/credential/presentation/04-send-authorization-response.ts
Comment (grausof) [inline src/credential/presentation/04-send-authorization-response.ts:141]: I'll put it here for future reference. Better to use speaking variable names
```suggestion
  const rsaPublicJwk = chooseRSAPublicKeyToEncrypt(rpConf);
```

File: src/credential/presentation/04-send-authorization-response.ts
Comment (grausof) [inline src/credential/presentation/04-send-authorization-response.ts:120]: Is there a reference for this task?

File: src/credential/issuance/06-obtain-credential.ts
Comment (balanza) [reply src/credential/issuance/06-obtain-credential.ts]: Fortunately, it was just a mistake of us. Fixed in https://github.com/pagopa/io-react-native-wallet/pull/62/commits/c5c8c37d35381f3b7d83186018a04cfc4d8f0ff0

File: src/credential/issuance/06-obtain-credential.ts
Comment (balanza) [reply src/credential/issuance/06-obtain-credential.ts]: We don't really need commented code. Removed in https://github.com/pagopa/io-react-native-wallet/pull/62/commits/32063c845f6fbc0b5410eda0904733319569ca16

File: src/credential/issuance/04-complete-user-authorization.ts
Comment (balanza) [reply src/credential/issuance/04-complete-user-authorization.ts]: I agree. I removed this function and implemented _no-op_ in the example app in https://github.com/pagopa/io-react-native-wallet/pull/62/commits/eb3a27676c9120b747d2256766b8479fb2a2c23d

File: src/utils/par.ts
Comment (balanza) [reply src/utils/par.ts]: I rearranged the code in https://github.com/pagopa/io-react-native-wallet/pull/62/commits/454042f6f2351d5716c7a0c2c4e2377838513c98, I guess it makes more sense now

File: src/credential/presentation/04-send-authorization-response.ts
Comment (balanza) [reply src/credential/presentation/04-send-authorization-response.ts]: It was actually a wring comment, removed in https://github.com/pagopa/io-react-native-wallet/pull/62/commits/42caeb6fcedc07aa5990371f87c80d3efdb3c8ea

File: src/trust/types.ts
Comment (balanza) [reply src/trust/types.ts:145]: done in https://github.com/pagopa/io-react-native-wallet/pull/62/commits/42caeb6fcedc07aa5990371f87c80d3efdb3c8ea

**Comment by balanza** [comment]
> Is it possible to investigate what is due to this?

I couldn't actually replicate the issue. I'm executing the app on a physical Android device (OnePlus 6T)

**Comment by grausof** [comment]
> > Is it possible to investigate what is due to this?
> 
> I couldn't actually replicate the issue. I'm executing the app on a physical Android device (OnePlus 6T)

After the latest commits it seems that the problem has disappeared!

**Review by grausof — APPROVED** [review]
Really great work 🚀 
P.S. Given the size of the PR I would also wait for another approve from @hevelius before the merge.
☝ **Remember to change the package version!**

**Review by hevelius — APPROVED** [review]
LGTM. Tested on simulator without errors. Great job!

**Pull request merged**
